harness/determined/pytorch/_lightning/adapter.py

```diff
@@ -0,0 +1,184 @@
+            lm.on_train_epoch_start()
+
+        def on_validation_epoch_start() -> None:
+            lm.on_validation_epoch_start()
+
+        cb.on_validation_epoch_start = on_validation_epoch_start
+        cb.on_validation_epoch_end = on_validation_epoch_end
+        cb.on_train_epoch_start = on_train_epoch_start
+        cb.on_train_epoch_end = on_train_epoch_end
+        return {"_lightning_module": cb}
+
+    def setup_optimizers_schedulers(
+        self,
+        context: PyTorchTrialContext,
+    ) -> Tuple[List[Optimizer], List[_LRScheduler]]:
+        """
+        Wrap optimizers and lr_schedulers returned by `configure_optimizers` to
+        work with Determined.
+        Return: Wrapped `optimizers`, and `lr_schedulers` in a tuple
+        """
+        optimizers, lr_scheduler_dicts, opt_frequencies = TrainerOptimizersMixin().init_optimizers(
+            self.lm
+        )
+        for freq in opt_frequencies:
+            check.eq(freq, 1, "custom optimizer frequencies are not supported")
+        optimizers = cast(List[Optimizer], optimizers)
+        lr_scheduler_dicts = cast(List[dict], lr_scheduler_dicts)
+
+        def lightning_scheduler_dict_to_det(lrs: dict) -> _LRScheduler:
+            """
+            input_dict = {
+                'scheduler': None,
+                'name': None,  # no custom name
+                'interval': 'epoch',  # after epoch is over
+                'frequency': 1,  # every epoch/batch
+                'reduce_on_plateau': False,  # most often not ReduceLROnPlateau scheduler
+                'monitor': monitor,  # value to monitor for ReduceLROnPlateau
+                'strict': True,  # enforce that the monitor exists for ReduceLROnPlateau
+            }
+            """
+            # TODO(DET-5021) support custom frequencies with the manual step.
+            step_mode = (
+                LRScheduler.StepMode.STEP_EVERY_EPOCH
+                if lrs["interval"] == "epoch"
+                else LRScheduler.StepMode.STEP_EVERY_BATCH
+            )
+            return context.wrap_lr_scheduler(lrs["scheduler"], step_mode)
+
+        optimizers = [self.context.wrap_optimizer(opt) for opt in optimizers]
+        lr_schedulers = [lightning_scheduler_dict_to_det(lrs) for lrs in lr_scheduler_dicts]
+        return optimizers, lr_schedulers
+
+    def train_batch(
+        self, batch: TorchData, epoch_idx: int, batch_idx: int
+    ) -> Union[torch.Tensor, Dict[str, Any]]:
+        self.lm.on_train_batch_start(batch, batch_idx, dataloader_idx=0)
+        type(self.lm).global_step = batch_idx
+
+        opt_metrics = []
+
+        for opt_idx, opt in enumerate(self.optimizers):
+            with monkey_patch(self.lm, "optimizers", lambda *args, **kwargs: self.optimizers):
+                self.lm.toggle_optimizer(opt, opt_idx)
+            metrics = self.lm.training_step(batch, batch_idx, optimizer_idx=opt_idx)
+
+            self.lm.on_train_batch_end(metrics, batch, batch_idx, dataloader_idx=0)
```

Comment:
Should this be called after the backward pass?

Reply:
yeah you're right https://github.com/PyTorchLightning/pytorch-lightning/blob/6cc1a060785448eacd016f6689d592ac822adaaf/pytorch_lightning/trainer/training_loop.py#L505